backup/moodle2/restore_format_vsf_plugin.class.php

```diff
@@ -106,20 +131,100 @@ public function after_restore_course() {
                     $DB->execute("UPDATE {course_sections} SET visible = 0 WHERE course = ? AND section = ?",
-                        [$this->step->get_task()->get_courseid(), $sectionnum]);
+                            [$this->step->get_task()->get_courseid(), $sectionnum]);
                 }
             }
         }
     }
+
+    /**
+     * Creates a dummy path element in order to be able to execute code after restore
+     *
+     * @return restore_path_element[]
+     */
+    public function define_module_plugin_structure() {
+        // Since this method is executed before the restore we can do some pre-checks here.
+        // In case of merging backup into existing course find the current number of sections.
+        $target = $this->step->get_task()->get_target();
+
+        $paths = [];
+        // We will ONLY perform the specifics if we're NOT importing etc etc.
+        $allowrestore = [backup::TARGET_NEW_COURSE, backup::TARGET_CURRENT_ADDING, backup::TARGET_EXISTING_ADDING];
+        if (in_array($target, $allowrestore)) {
+            // We'll restore :).
+            $elename = 'modicon'; // This defines the postfix of 'process_*' below.
+            $elepath = $this->get_pathfor('/modicon');
+            $paths[] = new restore_path_element($elename, $elepath);
+        }
+
+        return $paths;
+    }
+
+    /**
+     * Process course level icon customisations.
+     */
+    public function process_courseicon($data) {
+        if (!is_object($data)) {
+            $data = (object) $data;
+        }
+        $modname = $data->name;
+        $filearea = "modicon_{$modname}";
+        $itemid = 0;
+        $mappingitemname = null;
+        $filesctxid = $this->task->get_old_contextid();
+        $this->add_related_files('format_vsf', $filearea, $mappingitemname, $filesctxid, $itemid);
+    }
+
+    /**
+     * Process module level icon customisations.
+     */
+    public function process_modicon($data) {
+        if (!is_object($data)) {
+            $data = (object) $data;
+        }
+        // Data, as stated in the backup class, holds the itemid, module name and original context.
+        // The reason for the original context is because at this stage, we DO NOT have
+        // an original context.
+        // It is due to $this->task->get_old_contextid() returning 0, consequently failing $this->add_related_files().
+        // The only way to get this right now, is to hook into "after_restore_course()".
+        // We can achieve this by filling a STATIC property (we CANNOT use an instance property, it WILL fail) with the data object.
+        // Then, in "after_restore_course()", we can simply call "add_related_files()" with the correct data.
+
+        // Add data object to static property for use later (see below).
+        static::$modicons[] = $data;
+    }
+
+    /**
+     * Holds data objects that refer to custom module instance icons.
+     *
+     * @var array
+     */
+    protected static $modicons = [];
```

Comment:
Should this be defined at the top of the class?  For maintainability and readability.